tests/server/metrics-unit.js

```diff
@@ -0,0 +1,231 @@
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain';
+      }
+
+      return 'foo';
+    });
+    test.mocks.request.body = 'bar';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 1);
+    assert.lengthOf(test.mocks.logger.error.getCall(0).args, 1);
+    assert.instanceOf(test.mocks.logger.error.getCall(0).args[0], Error);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 0);
+    assert.equal(test.mocks.statsdCollector.write.callCount, 0);
+    assert.equal(test.mocks.gaCollector.write.callCount, 0);
+
+    tearDown();
+  };
+
+  suite['Content-Type is text/plain, data is valid JSON, user is sampled'] = function () {
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain';
+      }
+
+      return 'wibble';
+    });
+    test.mocks.request.body = '{"foo":"bar","isSampledUser":true}';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 0);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 1);
+    var data = test.mocks.metricsCollector.write.getCall(0).args[0];
+    assert.lengthOf(Object.keys(data), 3);
+    assert.equal(data.agent, 'wibble');
+    assert.equal(data.foo, 'bar');
+    assert.equal(data.isSampledUser, true);
+
+    assert.equal(test.mocks.statsdCollector.write.callCount, 1);
+    assert.equal(test.mocks.statsdCollector.write.getCall(0).args[0], data);
+
+    assert.equal(test.mocks.gaCollector.write.callCount, 1);
+    assert.equal(test.mocks.gaCollector.write.getCall(0).args[0], data);
+
+    tearDown();
+  };
+
+  suite['Content-Type is text/plain;charset=UTF-8, data is valid JSON, user is sampled'] = function () {
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain;charset=UTF-8';
+      }
+
+      return 'foo';
+    });
+    test.mocks.request.body = '{"isSampledUser":true}';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 0);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 1);
+    var data = test.mocks.metricsCollector.write.getCall(0).args[0];
+    assert.lengthOf(Object.keys(data), 2);
+    assert.equal(data.agent, 'foo');
+    assert.equal(data.isSampledUser, true);
+
+    assert.equal(test.mocks.statsdCollector.write.callCount, 1);
+    assert.equal(test.mocks.statsdCollector.write.getCall(0).args[0], data);
+
+    assert.equal(test.mocks.gaCollector.write.callCount, 1);
+    assert.equal(test.mocks.gaCollector.write.getCall(0).args[0], data);
+
+    tearDown();
+  };
+
+  registerSuite(suite);
+
+  function setUp (requestGet) {
```

Comment:
> @shane-tomlinson: I don't care if you use a named function before it is defined.

I usually prefer to define helper functions at the bottom of a file instead of the top. I personally prefer it this way.